tests: add ABI check

henryiii opened the pull request:
<!--
Title (above): please place [branch_name] at the beginning if you are targeting a branch other than master. *Do not target stable*.
It is recommended to use conventional commit format, see conventionalcommits.org, but not required.
-->
## Description

<!-- Include relevant issues or PRs here, describe what changed and why -->

Adding tests for ABI. Just a very basic implementation at the moment. Didn't add it the tests because we don't run nox on most jobs. Maybe could move it to pytest-virtualenv.

henryiii committed:
tests: add ABI check

Signed-off-by: Henry Schreiner <henryschreineriii@gmail.com>

henryiii committed:
tests: use pytest-virtualenv to test abi

Signed-off-by: Henry Schreiner <henryschreineriii@gmail.com>

henryiii committed:
ci: enable abi test on some platforms

Signed-off-by: Henry Schreiner <henryschreineriii@gmail.com>

Lalaland commented:
One suggestion: Have this test assert that the sizeof()s all the major structures are the same

So internals, type_info, instance, etc.

Even better would be to compare the raw bytes and verify those are equivalent. 

(Simply write the bytes of interals to a file or something and then compare the files).

henryiii commented:
Ahh, can't do an ABI check against and older version of pybind11 and Python 3.11, duh. :)

```
/tmp/tmpco8vd2na/.env/lib/python3.11/site-packages/pybind11/include/pybind11/detail/type_caster_base.h:473:36: error: invalid use of incomplete type ‘PyFrameObject’ {aka ‘struct _frame’}
[232](https://github.com/pybind/pybind11/actions/runs/3316298068/jobs/5477937033#step:25:233)
        473 |                 "  " + handle(frame->f_code->co_filename).cast<std::string>() +
[233](https://github.com/pybind/pybind11/actions/runs/3316298068/jobs/5477937033#step:25:234)
            |                                    ^~
```

> One suggestion

Hmm, interesting, okay, I'll try to add that soon. Won't have too much time for a while, though.

henryiii commented on .github/workflows/ci.yml:
```diff
@@ -176,6 +176,10 @@ jobs:
 
+    # This tests ABI compatibility
+    - name: ABI test
+      run: pytest tests/extra_abi
```
```suggestion
      if: matrix.python != '3.11-dev'
      run: pytest tests/extra_abi
```

henryiii committed:
tests: modernize cross-version ABI check

Replace pytest-virtualenv with a plain venv, pin the expected
PYBIND11_INTERNALS_VERSION so a bump cannot pass CI unnoticed, and
update the baseline release to 3.0.4. Since master already moved to
internals v12 and no release has it, the interop check currently
asserts clean isolation instead of compatibility.

Assisted-by: ClaudeCode:claude-fable-5
